python/launcher.py

```diff
@@ -0,0 +1,121 @@
+        '''
+        Checks whether the component is registered with registry.
+        Returns the component if registered.
+        Returns None if not registered.
+        '''
+        # timout to rule out network delay
+        timeout = time.time() + network_delay
+        component = None
+        while time.time() <= timeout:
+            if componenet_service.is_registered_in_registry.is_set():
+                component = self.__component_service_registry_manager.\
+                    find_all_by_category(component_service_name)
+                self.__logger.debug(f'{component} is found.')
+                break
+            else:
+                time.sleep(0.001)  # do not hog CPU
+                continue
+        if component is None:
+            self.__logger.critical(f'TIMEOUT! {componenet_service} '
+                                   f'can not be found in registry.')
+        return component
+
+    def launch(self, actions):
+        '''
+        launches the necessary components (Application Companions and
+        Orchestrator) as per actions.
+        '''
+        application_companions = []
+        # i. initialize Application Comanions
+        for action in actions:
+            application_companions.append(ApplicationCompanion(
+                                    self._log_settings,
+                                    self._configurations_manager,
+                                    action,
+                                    self.__component_service_registry_manager))
+        # ii. launch the application companions
+        for application_companion in application_companions:
+            self.__logger.info('setting up Application Companion.')
+            application_companion.start()
+        # allow application companions to register with registry
+        time.sleep(0.5)  # TODO: set as run-time network delay
+
+        # iii. initialize steering service
+        self.__logger.info('setting up C&S service.')
+        steering_service = CommandSteeringService(
+                                    self._log_settings,
+                                    self._configurations_manager,
+                                    self.__component_service_registry_manager)
+        # iv. launch C&S service
+        steering_service.start()
+        # v. check if C&S is registered
+        if self.__is_component_registered(
+                    steering_service,
+                    SERVICE_COMPONENT_CATEGORY.COMMAND_AND_SERVICE) is None:
+            self.__logger.critical('command and steering service is \
+                                   not found in registry.')
+            return Response.ERROR
+
+        # v. initialize orchestrator
+        orchestrator = Orchestrator(self._log_settings,
+                                    self._configurations_manager,
+                                    self.__component_service_registry_manager)
+        # vi. launch orchestrator
+        orchestrator.start()
```

Comment:
The class in orchestrator.py implements the function run().
Check for correctness?

Reply:
Orchestrator, Application_Companions and Steering Service are separate processes. We override run() method of multiprocessing to implement the functionality to be executed by a new process. However, it is not recommended to call it explicitly rather we call start() to arrange for the run() method to be invoked in a separate process. For more details: https://docs.python.org/3/library/multiprocessing.html